crates/rethnet_evm/benches/state.rs

```diff
@@ -45,11 +45,106 @@ fn bench_code_by_hash(c: &mut Criterion) {
     );
 }
 
+fn bench_database_commit(c: &mut Criterion) {
+    use hashbrown::HashMap;
+    use revm::primitives::KECCAK_EMPTY;
+
+    use rethnet_eth::remote::optional_u64_from_hex;
+    use rethnet_evm::{Account, AccountInfo, StorageSlot};
+
+    #[derive(Debug, serde::Deserialize)]
+    #[serde(rename_all = "camelCase")]
+    struct AccountState {
+        #[serde(deserialize_with = "optional_u64_from_hex")]
+        nonce: Option<u64>,
+        balance: Option<U256>,
+        storage: HashMap<U256, Option<U256>>,
+        code: Option<Bytes>,
+        storage_cleared: bool,
+    }
+
+    let mut accounts_to_commit: HashMap<Address, Account> = HashMap::new();
+    let json_accounts: HashMap<Address, AccountState> = serde_json::from_str(
+        &std::fs::read_to_string(
+            "benches/fixtures/accounts_changed_in_mainnet_block_17295357.json",
+        )
+        .unwrap(),
+        /* a fresh set of account updates can be retrieved via, eg:
+            BLOCK=17295357 \
+                HARDHAT_EXPERIMENTAL_VM_MODE=ethereumjs \
+                HARDHAT_RUN_FULL_BLOCK_DUMP_STATE_TO_FILE=../../crates/rethnet_evm/benches/fixtures/accounts_changed_in_mainnet_block_${BLOCK}.json \
+                sh -c 'yarn ts-node scripts/test-run-forked-block.ts $ALCHEMY_URL $BLOCK'
+            note that this should be done from the packages/hardhat-core directory.
+        */
+    )
+    .unwrap();
+    for (address, account_state) in json_accounts.iter() {
+        let mut storage: HashMap<U256, StorageSlot> = HashMap::new();
+        for (location, value) in account_state.storage.clone() {
+            storage.insert(
+                location,
+                StorageSlot {
+                    original_value: U256::ZERO, // TODO: something better?
+                    present_value: value.unwrap_or(U256::ZERO),
+                },
+            );
+        }
+        let code = account_state
+            .code
+            .clone()
+            .map(|bytes| Bytecode::new_raw(bytes));
+        accounts_to_commit.insert(
+            *address,
+            Account {
+                info: AccountInfo {
+                    balance: account_state.balance.unwrap(),
+                    nonce: account_state.nonce.unwrap(),
+                    code: code.clone(),
+                    code_hash: code.map_or(KECCAK_EMPTY, |code| code.hash()),
+                },
+                storage,
+                storage_cleared: account_state.storage_cleared,
+                is_destroyed: false,
```

Comment:
Does your script also generate deletions of accounts (i.e. self-destruct)?

Reply:
the script just captures whatever transactions happened in a particular block. i haven't delved into what types of transactions are present in the block i picked. (i just picked an arbitrary recent one.) do you think it's important to have self-destruct transactions? if so, we could dig a bit to find such a block and use that one as the fixture instead of what's there now.

Reply:
The idea of this benchmark is for the `fn commit` is to handle a variety of changes that matches real-world use cases, hence why we capture and replay the block. An ideal block would be one that has some of all cases:
- self-destruct account
- insert account
- modify accounts
- insert storage slots
- modify storage slots
- remove storage slots

Naturally, it'll be hard to find a block that matches all of these, so potentially replaying a sequence of blocks would do the trick?

Were there any qualifying characteristics for the block you chose?

Reply:
No, there were no qualifying characteristics for the block I chose. I asked the team [here](https://nomicfoundation.slack.com/archives/CTTF55XPT/p1687302433214939)

Reply:
Given that the expectation is that it'll be tricky to find one block to rule them all, I'd go with the approach of selecting a range of blocks to replay. At least statistically you're more likely to get a little of everything.